### DISCUSSION TITLE
How to apply color map for a thresholded unstructured grid?

### DISCUSSION THREAD

**Discussion by tomek1911** [pyvista/pyvista#3547] [Q&A]
I want to apply a colour map to a mesh. The colour map is based on matplotlib ListedColormap - I call it slicer_map. My data is a segmentation volume of teeth with values from 1 to 32 (tooth IDs) and 0 which is background. I threshold with values above 1 because otherwise, I obtain a black cuboid. 
I have a problem because when I plot mesh it is all black, but cmap and colorbar seems to be working well. I tried to use label values as scalars but they are somehow ignored.

```
volume = np.load('volume_label.npy')
data = pv.wrap(volume)
outline = data.outline()
tresh_data = data.threshold(1)
p = pv.Plotter()
p.add_mesh(outline, color="k")
p.add_mesh(tresh_data, cmap=slicer_map, scalars="values")
p.show(screenshot="cbct3d.png")
```
<img width="759" alt="black mesh" src="https://user-images.githubusercontent.com/23128356/199834300-37198784-f1b8-4cc7-8233-bf0ccea21375.png">

The method below does the trick - but I want to avoid opacity. I want the mesh to be solid and later I want to visualise this model voxelised.
```
import pyvista as pv
import numpy as np

volume = np.load('volume_label.npy')
data = pv.wrap(volume)
data_inverted = data.threshold(value=1, invert=True)
data_inverted.plot(cmap=slicer_map, background='gray',
          opacity=[0]+[1 for i in range(32)],
          screenshot='cbct3d_2.png',
          zoom = 1.5, cpos = [1, -10, 1], window_size=[2048, 1024])
```
<img width="1017" alt="opacity_model" src="https://user-images.githubusercontent.com/23128356/199834146-55a0ea78-778d-463e-912c-b32b2227435a.png">

Am I missing something, is there any other method to colourise mesh/voxels faces?

my numpy array: 
[volume_label.npy.zip](https://github.com/pyvista/pyvista/files/9933038/volume_label.npy.zip)

**Comment by banesullivan** [comment]
You may want to switch this mesh so that the values lie on the CELLs of the mesh rather than the POINTs. `pv.wrap()` of a NumPy array like this assumes POINT data but you may actually want CELL data - such that each element in the 3D numpy array represents a voxel in the PyVista `UniformGrid`

For instance:

```py
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

volume = np.load('volume_label.npy')
data = pv.UniformGrid()
data.dimensions = np.array(volume.shape) + 1
data.cell_data['values'] = volume.ravel(order='F')
outline = data.outline()

tresh_data = data.threshold(1, scalars='values')

p = pv.Plotter(notebook=0)
p.add_mesh(outline, color="k")
p.add_mesh(tresh_data, cmap='tab20', scalars="values")
p.show()
```

<img width="624" alt="Screen Shot 2022-11-03 at 7 02 39 PM" src="https://user-images.githubusercontent.com/22067021/199863001-8ad90c67-8a26-43ad-b2f4-023e17aef242.png">

**Reply by banesullivan** [reply]
These resource might help if you are unfamiliar:

- https://docs.pyvista.org/user-guide/data_model.html#data-arrays (see point data and cell data subsections)
- https://docs.pyvista.org/examples/00-load/create-uniform-grid.html

**Reply by tomek1911** [reply]
Thank you so much for your help. That's exactly what I needed. Yesterday I started reading docs from scratch to understand the basics, and I got suspicious about cell data. 

I still need to figure out why the colour map doesn't fit values from the array, but when I plot in pyplot colours are assigned properly. 
The tooth marked with a red arrow should have a value 2, not 0.
<img width="1200" alt="Screenshot 2022-11-04 at 12 29 17" src="https://user-images.githubusercontent.com/23128356/199962952-d01c9f93-13ea-4ca5-a614-c79f173ce87c.png">
Proper colormap:
![labels](https://user-images.githubusercontent.com/23128356/199963354-c1f0e70b-1bcf-4239-aa1d-2c778903766a.png)

I hope I will work it out on my own ;) if I do I will write the solution here.

**Reply by banesullivan** [reply]
> The tooth marked with a red arrow should have a value 2, not 0.

It does have a value of two! Take a look at the colormap... black is currently mapped to 2.

The issue is that PyVista is not handling your colormap how you expect. We need to do a bit of shifting of the `clim` in order for the colors to be properly mapped. After @akaszynski's recent working wrapping `LookupTables` this should be quite easy.

Would you do me a favor and please share the code to generate you matplotlib ListedColormap so that I can put together an example for you?

**Reply by akaszynski** [reply]
BTW, this is a super cool dataset. Can we use it an example somehow?

**Reply by tomek1911** [reply]
Hi, sorry, but I missed your message. Could you please explain more precisely what you need?

**Reply by banesullivan** [reply]
Could you share the code to make your colormap? the `slicer_map` object

**Reply by tomek1911** [reply]
The colours I use are the default colour map from 3D Slicer - I exported the colours (RGBA) from the program to the CSV file [colormap_slicer.csv](https://github.com/pyvista/pyvista/files/10057777/colormap_slicer.csv).


<img width="496" alt="Screenshot 2022-11-16 at 22 08 06" src="https://user-images.githubusercontent.com/23128356/202295133-778c1fbc-310d-4a42-8205-b396ba2223e6.png">


```python
import pandas as pd
from matplotlib import colors

def get_colormap(file_path : str ='csv_files/colormap_slicer.csv') -> colors.ListedColormap:
      tooth_colors = pd.read_csv(file_path, delimiter=";", header=None)
      tooth_colors_df = tooth_colors.iloc[:,2:5]
      tooth_colors_df.columns = ['r','g','b']
      slicer_colorspace = tooth_colors_df.to_numpy()/255
      slicer_map = colors.ListedColormap(slicer_colorspace, 'slicer_colors')
      return slicer_map
```

**Reply by banesullivan** [reply]
Would you please attach that csv file?

**Reply by banesullivan** [reply]
Sorry for the delay -- I did not get a notification of you adding the CSV file to your previous comment.

All you need to do is pass that cmap to PyVsita and set the `clim` when plotting to properly map the values

I also recommend looking at:

- https://docs.pyvista.org/examples/02-plot/scalar-bars.html#sphx-glr-examples-02-plot-scalar-bars-py
- https://docs.pyvista.org/examples/02-plot/lookup-table.html#lookup-tables

```py
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt
import pandas as pd
from matplotlib import colors

volume = np.load('volume_label.npy')
data = pv.UniformGrid()
data.dimensions = np.array(volume.shape) + 1
data.cell_data['values'] = volume.ravel(order='F')
outline = data.outline()

tresh_data = data.threshold(1, scalars='values')

def get_colormap(file_path : str ='colormap_slicer.csv') -> colors.ListedColormap:
    tooth_colors = pd.read_csv(file_path, delimiter=";", header=None)
    tooth_colors_df = tooth_colors.iloc[:,2:5]
    tooth_colors_df.columns = ['r','g','b']
    slicer_colorspace = tooth_colors_df.to_numpy()/255
    slicer_map = colors.ListedColormap(slicer_colorspace, 'slicer_colors')
    return slicer_map

cmap = get_colormap()

p = pv.Plotter(notebook=0)
p.add_mesh(outline, color="k")
p.add_mesh(tresh_data, cmap=cmap, scalars="values", clim=[-.5, 32.5])
p.show()
```
<img width="624" alt="Screen Shot 2022-12-30 at 10 40 01 AM" src="https://user-images.githubusercontent.com/22067021/210098159-d10f94d9-cda1-4224-94dd-2b98d8e43ba8.png">
